sklearn/datasets/_openml.py

```diff
@@ -311,6 +333,9 @@ def _convert_arff_data_dataframe(arff, columns, features_dict):
     attributes = OrderedDict(arff['attributes'])
     arff_columns = list(attributes)
 
+    if isinstance(arff['data'], tuple):
+        raise ValueError("Unreachable code. arff['data'] must be a generator.")
+
     # calculate chunksize
     first_row = next(arff['data'])
```

Comment:
The above check is necessary as technically `arff` could be a tuple for sparse data, and then `next` would fails as it's not implemented for tuples.

Reply:
Can we directly check that `arff['data'` is not iterable?

```py
if not isinstance(arff['data'], Iterable)
```

(Not for this PR: This reminds me that we can likely return dataframes with sparse extension arrays)

Reply:
Yes, done.